presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -1084,26 +1085,32 @@ private GroupedExecutionProperties processChildren(PlanNode node)
             //     * if all children is "recovery eligible", this node is "recovery eligible"
             //     * if any child is "not recovery eligible", this node is "not recovery eligible"
             boolean anyUseful = false;
-            OptionalInt totalLifespans = OptionalInt.empty();
+            OptionalInt totalLifespansOptional = OptionalInt.empty();
             boolean allRecoveryEligible = true;
-            ImmutableList.Builder<PlanNodeId> capableTableScanNodes = ImmutableList.builder();
+            ImmutableList.Builder<PlanNodeId> capableTableScanNodesBuilder = ImmutableList.builder();
+
             for (PlanNode source : node.getSources()) {
                 GroupedExecutionProperties properties = source.accept(this, null);
                 if (!properties.isCurrentNodeCapable()) {
                     return GroupedExecutionProperties.notCapable();
                 }
                 anyUseful |= properties.isSubTreeUseful();
                 allRecoveryEligible &= properties.isRecoveryEligible();
-                if (!totalLifespans.isPresent()) {
-                    totalLifespans = OptionalInt.of(properties.totalLifespans);
+                if (!totalLifespansOptional.isPresent()) {
+                    totalLifespansOptional = OptionalInt.of(properties.totalLifespans);
                 }
                 else {
-                    checkState(totalLifespans.getAsInt() == properties.totalLifespans, format("Mismatched number of lifespans among children nodes. Expected: %s, actual: %s", totalLifespans.getAsInt(), properties.totalLifespans));
+                    checkState(totalLifespansOptional.getAsInt() == properties.totalLifespans, format("Mismatched number of lifespans among children nodes. Expected: %s, actual: %s", totalLifespansOptional.getAsInt(), properties.totalLifespans));
                 }
+                capableTableScanNodesBuilder.addAll(properties.capableTableScanNodes);
+            }
 
-                capableTableScanNodes.addAll(properties.capableTableScanNodes);
+            int totalLifespans = totalLifespansOptional.getAsInt();
+            List<PlanNodeId> capableTableScanNodes = capableTableScanNodesBuilder.build();
+            if (capableTableScanNodes.size() * totalLifespans > getMaxBucketsForGroupedExecution(session)) {
```

Comment:
`totalLifespans` only consider the number of buckets in the table right? -- But what usually contribute to memory is `#buckets * #partition`?

Reply:
Actually yes. It feels like it should be checked in the connector. The problem is that currently it is not possible to disable grouped execution from the connector. The only option is to remove the bucketing handle all together, that would effectively remove the `$bucket` virtual column and the partitioning as well.

Reply:
Yeah, i can remove bucketing entirely for these queries, but it will make queries on the $bucket column fail, and also means we won't be able to take advantage of bucketing for colocated joins.  but nowhere else has info about number of partitions.

Reply:
The easiest solution would be to add a method to the connector API.  something like canScheduleGroupedExecution().  and then we could have the connector make the decision without removing bucketing.  But that would be yet another api change. I'll keep thinking about whether there's a better way.

Reply:
I was thinking to have some extra information (e.g. `TableHandle` or `TableLayoutHandle`? ) into `ConnectorNodePartitioningProvider#listPartitionHandles`: https://github.com/prestodb/presto/blob/7518a70d01edd99de9ebdc7e8cedefb11b298563/presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorNodePartitioningProvider.java#L38-L41

So `HiveNodePartitioningProvider` can look at whether it's reading from too many partitions * buckets, if so, it returns `singletonList(NOT_PARTITIONED)`, claim it doesn't support per-bucket split discovery.

Reply:
@arhimondr 

> The problem is that currently it is not possible to disable grouped execution from the connector.

It's possible, the connector can claim the table is bucketed, but I don't support per-bucket split discovery. For, example, Raptor connector :D 

However, now I rethink about it, maybe Connector should always support per-bucket split discovery for bucket table in a long term. 

So I am open to "virtually" remove table bucket property if it scans too many partitions. As @rongrong points out, we anyway want to penalize queries don't have any partition filter.  Using `$bucket` filter but not use partition filter sounds like user is optimizing the query in the wrong direction.

Reply:
Ok, so it sounds like we're converging on Hive returning that the table isn't bucketed if it queries too many buckets + partitions.  That could be a confusing user experience if the $bucket column is there sometimes, but not others, so maybe we should issue a warning too.